bitbybit/src/bitenum.rs

```diff
@@ -1,303 +1,225 @@
+            max_discr = value;
+            max_discr_span = discriminant.__span();
+        }
+    }
+    if max_discr >= max_count {
+        let max_value = max_count - 1;
+        let msg = format!("{DISCR_TOO_LARGE}, max value is {max_value}");
+        return Err(syn::Error::new(max_discr_span, msg));
+    }
+    Ok(())
+}
 
-    // There are two ways to turn an int into an enum values:
-    // - match cases against every single integer
-    // - exclude unhandled integers, followed by transmute (unsafe)
-    // For now, we'll go with the first option. If we find that the compiler generates bad code,
-    // we can switch to the second option (as we required all values to be literals, so we can
-    // analyze used vs unused ranges)
+pub(crate) fn bitenum(config: Config, input: syn::ItemEnum) -> syn::Result<TokenStream> {
+    let config = config.explicit()?;
+    check_explicit_conditional(&config, &input)?;
+    check_explicit_exhaustive(&config, &input)?;
 
-    let case_values: Vec<TokenStream2> = emitted_variants
-        .iter()
-        .map(|(expression, _int_value, name, cfg_attributes)| {
-            if return_is_result {
-                quote! {
-                    #( #cfg_attributes )*
-                    #expression => Ok(Self::#name),
-                }
-            } else {
-                quote! {
-                    #( #cfg_attributes )*
-                    #expression => Self::#name,
-                }
-            }
-        })
-        .collect();
+    let bits = config.bits;
+    let (base_type, qualified_type) = (bits.base_type()?, bits.qualified_path()?);
+    let (constructor, reader) = (bits.constructor()?, bits.reader());
 
-    let constructor_function = if return_is_result {
-        quote!(
-            /// Creates a new instance of this bitfield with the given raw value, or
-            /// Err(value) if the value does not exist in the enum.
-            pub const fn new_with_raw_value(value: #bounded_data_type) -> Result<Self, #base_data_type> {
-                match value #bounded_getter {
-                    #( #case_values )*
-                    _ => Err(value #bounded_getter)
-                }
-            }
-        )
-    } else {
-        let panic_string = format!("{}: Unhandled value", enum_name);
-        quote!(
-            /// Creates a new instance of this bitfield with the given raw value.
-            ///
-            /// As the enum is exhaustive, this function will always return a valid result
-            pub const fn new_with_raw_value(value: #bounded_data_type) -> Self {
-                match value #bounded_getter {
-                    #( #case_values )*
-                    _ => panic!(#panic_string)
-                }
-            }
-        )
+    let non_exhaustive = config.exhaustive.matches(false);
+    let ok = non_exhaustive.then_some(quote!(Ok));
+    let new_return_type = match non_exhaustive {
+        true => quote!(std::result::Result<Self, #base_type>),
```

Comment:
fyi, I had an outstanding patch to fix the Err-Type here: https://github.com/danlehmann/bitfield/pull/47

But as that's a breaking change, I held off from doing this one until a major version bump